sim/src/main/cc/bridges/TracerVModule.cc

```diff
@@ -0,0 +1,242 @@
+
+    // the value of the first cycle as returned from TracerV
+    const uint64_t cycle_offset = 3;
+
+    // modified as we go
+    uint64_t e_cycle = cycle_offset;
+
+    const unsigned tracerv_width = tracerv.max_core_ipc;
+
+    // load MMIO and capture expected outputs
+    auto load = [&](std::vector<uint64_t> iad, std::vector<bool> bit) {
+      std::vector<uint64_t> valid_i;
+      assert(iad.size() == bit.size());
+      for (unsigned i = 0; i < iad.size(); i++) {
+        // std::cout << "loading " << i << " with " << iad[i] << "," << bit[i]
+        // << std::endl;
+        poke(namei(i), iad[i], true);
+        poke(namev(i), bit[i], true);
+
+        // calculate what TraverV should output, and save it
+        if (bit[i]) {
+          valid_i.emplace_back(iad[i]);
+        }
+      }
+
+      // place instructions onto the vector 7 at a time
+      for (size_t i = 0; i < valid_i.size(); i += 7) {
+        auto last = std::min(valid_i.size(), i + 7);
+        std::vector<uint64_t> chunk =
+            std::vector<uint64_t>(valid_i.begin() + i, valid_i.begin() + last);
+        expected_pair.emplace_back(std::make_pair(e_cycle, chunk));
+      }
+
+      e_cycle++;
+    };
+
+    // loop over tests. choose random valids with a simple pattern of iaddr
+    // load into MMIO, and tick the system
+    for (unsigned test_step = 0; test_step < get_total_trace_tests();
+         test_step++) {
+      const uint64_t pull = rand_next(tracerv_width + 1);
+
+      auto pull_iaddr = get_iaddrs(test_step, tracerv_width);
+      auto pull_bits = get_contiguous(pull, tracerv_width);
+
+      load(pull_iaddr, pull_bits);
+      steps(1);
+    }
+
+    const auto &[final_iaddr, final_valid] = get_final_values(tracerv_width);
+
+    // load final values (which are not valid and thus not checked)
+    load(final_iaddr, final_valid);
+
+    tracerv.flush();
+
+    steps(100);
+
+    // write out a file which contains the expected output
+    write_expected_file();
+  }
+
+  /**
+   * Writes the contents of the member variable expected_pair
+   * to the expected file pointer. This is done via the tracerv_t::serialize
+   * function.
+   */
+  void write_expected_file() {
+    uint64_t OUTBUF[8];
+
+    for (const auto &beat : expected_pair) {
+      const auto &[cycle, insns] = beat;
+      memset(OUTBUF, 0, sizeof(OUTBUF));
+      OUTBUF[0] = cycle;
+      assert(insns.size() < 8);
+      for (int i = 0; i < insns.size(); i++) {
+        OUTBUF[i + 1] = insns[i] | tracerv_t::valid_mask;
+      }
+
+      // because expected_pair doesn't contain the instruction value for non
+      // valid instructions only the human readable output will compare
+      // correctly
+      tracerv_t::serialize(OUTBUF,
+                           sizeof(OUTBUF),
+                           expected,
+                           nullptr,
+                           tracerv.max_core_ipc,
+                           true,
+                           false,
```

Comment:
If the arguments are true/false or integral constants, please add a `/*argname=*/` comment before them.